File: java/pg-data-adapter/src/main/java/com/leecampbell/cod/data/PostgresqlRepository.java
package com.leecampbell.cod.data;

import com.leecampbell.cod.domain.contracts.DomainEvent;
import com.leecampbell.cod.domain.model.Loan;
import com.leecampbell.cod.domain.services.Repository;

import java.sql.SQLException;
import java.util.UUID;

public final class PostgresqlRepository implements Repository {
    private final EventAppender writer;
    private final EventReader reader;

    public PostgresqlRepository(String connectionUrl){
        this.reader = new EventReader(connectionUrl);
        this.writer = new EventAppender(connectionUrl);
    }

    @Override
    public Loan get(UUID id) {
        try {
            var currentState = reader.readStreamInstance("loan", id.toString());
            var loan = new Loan(id);
            for (var event : currentState) {
                loan.applyEvent(event);
            }
            return loan;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void save(Loan item) {
        try {
            var uncommittedEvents = item.getUncommittedEvents();
            int expectedVersion = item.getVersion() - uncommittedEvents.length;
            writer.append("loan", item.getId().toString(), expectedVersion, uncommittedEvents);
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }
}

File: typescript/packages/pg-data-adapter/src/postgresql-repository.ts
import { Pool } from "pg";
import { Loan, type Repository } from "@cod/domain";
import { EventReader } from "./event-reader.js";
import { EventAppender } from "./event-appender.js";

export class PostgresqlRepository implements Repository<Loan> {
  readonly #pool: Pool;
  readonly #reader: EventReader;
  readonly #writer: EventAppender;

  constructor(connectionString: string) {
    this.#pool = new Pool({ connectionString });
    this.#reader = new EventReader(this.#pool);
    this.#writer = new EventAppender(this.#pool);
  }

  async get(id: string): Promise<Loan> {
    const events = await this.#reader.readStreamInstance("loan", id);
    const loan = new Loan(id);
    for (const event of events) {
      loan.applyEvent(event);
    }
    return loan;
  }

  async save(loan: Loan): Promise<void> {
    const uncommittedEvents = loan.getUncommittedEvents();
    const expectedVersion = loan.version - uncommittedEvents.length;
    await this.#writer.append(
      "loan",
      loan.id,
      expectedVersion,
      uncommittedEvents,
    );
    loan.clearUncommittedEvents();
  }

  async healthCheck(): Promise<boolean> {
    try {
      await this.#pool.query("SELECT 1");
      return true;
    } catch {
      return false;
    }
  }

  async close(): Promise<void> {
    await this.#pool.end();
  }
}
